grammar.js

```diff
@@ -1,284 +1,391 @@
+      $.let_expression,
+      $.function_call,
+      $.index_expression,
+      $.dot_index_expression,
+      $.assert_expression,
+      $._literal,
+      $._variable_name,
+    ),
+    let_expression: ($) => bodied_block('let', $.parenthesized_assignments, $._expression),
 
-  // atoms that create immediate values
-  _literal: $ => choice(
-    $.string,
-    $.number,
-    $.boolean,
-    $.undef,
-    $.function,
-    $.range,
-    $.list,
-  ),
-  string: $ => token(seq('"', repeat(choice(/[^"]/, '\\"')), '"')),
-  number: $ => seq(/\d+\.\d*|\.\d+|\d+/, optional(/e-?\d+/)),
-  boolean: $ => choice('true', 'false'),
-  undef: $ => 'undef',
+    // valid names for variables, functions, and modules
+    identifier: () => /[a-zA-Z_]\w*/,
+    special_variable: () => token(seq('$', /[a-zA-Z_]\w*/)),
+    _variable_name: ($) => choice($.identifier, $.special_variable),
 
-  // compound atoms that are still literals
-  function: $ => seq(
-    'function',
-    field('parameters', $.parameters_declaration),
-    field('body', $._expression),
-  ),
-  range: $ => brackets(seq(
-    field('start', $._expression),
-    optional(seq(':', field('increment', $._expression))),
-    ':', field('end', $._expression),
-  )),
+    // atoms that create immediate values
+    _literal: ($) => choice(
+      $.string,
+      $.number,
+      $.boolean,
+      $.undef,
+      $.function,
+      $.range,
+      $.list,
+    ),
+    string: () => token(seq('"', repeat(choice(/[^"]/, '\\"')), '"')),
+    number: ($) => choice($.decimal, $.float),
+    decimal: () => token(/\d+/),
+    float: () => token(/-?(\d+(\.\d+)?|\.\d+)(e-?\d+)?/),
+    boolean: () => choice('true', 'false'),
+    undef: () => 'undef',
 
-  list: $ => brackets(seq(commaSep($._list_cell), optional(','))),
-  _list_cell: $ => choice($._expression, $.each, $.list_comprehension),
-  _comprehension_cell: $ => choice(
-    $._expression,
-    opt_grouping(parens, $.each),
-    opt_grouping(parens, $.list_comprehension)
-  ),
-  each: $ => seq('each', choice($._expression, $.list_comprehension)),
+    // compound atoms that are still literals
+    function: ($) => seq(
+      'function',
+      field('parameters', $.parameters_declaration),
+      field('body', $._expression),
+    ),
+    range: ($) => brackets(seq(
+      field('start', $._expression),
+      optional(seq(':', field('increment', $._expression))),
+      ':', field('end', $._expression),
+    )),
 
-  list_comprehension: $ => seq(
-    choice($.for_clause, $.if_clause)
-  ),
-  for_clause: $ => seq('for',
-    choice($.parenthesized_assignments, $.condition_update_clause),
-    $._comprehension_cell
-  ),
-  if_clause: $ => prec.right(seq(
-    'if',
-    field('condition', $.parenthesized_expression),
-    field('consequence', $._comprehension_cell),
-    optional(
-      seq('else', field('alternative', $._comprehension_cell))
-    )
-  )),
+    list: ($) => brackets(seq(commaSep($._list_cell), optional(','))),
```

Comment:
This `commaSep` invocation should probably now use the optional `trailing_separator` argument.